go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -356,7 +357,18 @@ func (ts *txThrottlerStateImpl) throttle() bool {
 	// Serialize calls to ts.throttle.Throttle()
 	ts.throttleMu.Lock()
 	defer ts.throttleMu.Unlock()
-	return ts.throttler.Throttle(0 /* threadId */) > 0
+
+	var maxLag uint32
+
+	for tabletType := range ts.tabletTypes {
+		maxLagPerTabletType := ts.throttler.LastMaxLagNotIgnoredForTabletType(tabletType)
+		if maxLagPerTabletType > maxLag {
+			maxLag = maxLagPerTabletType
+		}
+	}
```

Comment:
Is this required to be checked on every call? This probably can increase the query execution time.

Reply:
The decisions to throttle needs to be made on every call, so if we want to make it with up to date information, it would be required.

I guess we could cache it in memory for a bit and only update it every certain amount of time. Or only evaluate it if/when the first condition of the return value `ts.throttler.Throttle(0) > 0`  is `True`. Any preferences?

Reply:
As a general design opinion, the throttler should generate decisions offline and irrespective of any requests. Its decisions should have some expiration, ie this can be implemented via cache. Then, when a query comes by, it shouldn't really talk to the throttler: it should only read the already populated info in the cache.

Anyway, that's how I think it should be done. It certainly takes a lot of locking and care to implement it on a per-query basis.

Reply:
Those are my thoughts as well. Throttling decision should not be made on the hot path. A separate throttling goroutine should keep updating the decision.

Reply:
In that case, I will implement this as a cache.

Notice, however, that this will not completely remove the throttling code from the query path, as even before the change of this PR, or any of the code we have added to the throttler, it was already the case that the decision about whether to throttle or not depended on a call to check `ts.throttler.Throttle(0 /* threadId */) > 0` that would be checked on the hot path. And while I do see the advantage of having it separately to avoid a latency hit for queries, I think such a change would fall outside the scope of the bug that this PR aims to fix. What do you think?

Reply:
Agreed that it's outside the scope of this PR, and that a cache is the next best thing right now.